lib/Sema/AvailableFactsAnalysis.cpp

```diff
@@ -0,0 +1,458 @@
+//===-------- AvailableFactsAnalysis.cpp - collect comparison facts -------===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+//
+//  This file implements an analysis for collecting comparison facts.
+//
+//  The analysis has the following characteristics: 1. forward dataflow analysis,
+//  2. conservative, 3. intra-procedural, and 4. path-sensitive.
+//
+//===----------------------------------------------------------------------===//
+
+#include "clang/Sema/AvailableFactsAnalysis.h"
+
+namespace clang {
+class Sema;
+
+void AvailableFactsAnalysis::Analyze() {
+  assert(Cfg && "expected CFG to exist");
+
+  std::vector<Comparison> AllComparisons;
+  std::queue<ElevatedCFGBlock *> WorkList;
+  std::vector<ElevatedCFGBlock *> Blocks;
+
+  PostOrderCFGView POView = PostOrderCFGView(Cfg);
+  unsigned int MaxBlockID = 0;
+  std::vector<unsigned int> InWorkList;
+  for (const CFGBlock *Block : POView)
+    if (Block->getBlockID() > MaxBlockID)
+      MaxBlockID = Block->getBlockID();
+
+   BlockIDs.clear();
+   BlockIDs.resize(MaxBlockID + 1, -1);
+   for (const CFGBlock *Block : POView) {
+     auto NewBlock = new ElevatedCFGBlock(Block);
+     WorkList.push(NewBlock);
+     InWorkList.push_back(Block->getBlockID());
+     Blocks.push_back(NewBlock);
+     BlockIDs[Block->getBlockID()] = Blocks.size() - 1;
+   }
+
+   // Compute Gen Sets
+   for (auto B : Blocks) {
+     if (const Stmt *Term = B->Block->getTerminator()) {
+       if(const IfStmt *IS = dyn_cast<IfStmt>(Term)) {
+         ComparisonSet Comparisons;
+         ExtractComparisons(IS->getCond(), Comparisons);
+         B->GenThen.insert(Comparisons.begin(), Comparisons.end());
+
+         ComparisonSet NegatedComparisons;
+         ExtractNegatedComparisons(IS->getCond(), NegatedComparisons);
+         B->GenElse.insert(NegatedComparisons.begin(), NegatedComparisons.end());
+       }
+     }
+     AllComparisons.insert(AllComparisons.end(), B->GenThen.begin(), B->GenThen.end());
+     AllComparisons.insert(AllComparisons.end(), B->GenElse.begin(), B->GenElse.end());
+   }
+
+   // Which comparisons contain pointer derefs?
+   std::vector<bool> ComparisonContainsDeref;
+   for (auto C : AllComparisons) {
+     if (ContainsPointerDeref(C))
+       ComparisonContainsDeref.push_back(true);
+     else
+       ComparisonContainsDeref.push_back(false);
+   }
+
+   // Which blocks contain potential pointer assignments? (assignment via a pointer or a call)
+   std::vector<bool> PointerAssignmentInBlocks(Blocks.size(), false);
+   for (unsigned int Index = 0; Index < Blocks.size(); Index++) {
+     std::set<const VarDecl *> DefinedVars;
+     for (CFGElement Elem : *(Blocks[Index]->Block))
```

Comment:
This isn't the right test for a pointer assignment.   You should be checking for assignments to lvalues produced by a pointer-dereferencing expressions (this is the function f(e) described in my comment for checking for a pointer-dereference).   You should be checking for:

1. An assignment operator whose left-hand side is an expression e where f(e) is true.
2. An increment/decrement operator whose child expression  e is an expression for which f(e) is true.